tools/samtools.wdl

```diff
@@ -334,6 +388,101 @@ task subsample {
+    }
+
+    parameter_meta {
+        bam: "Input BAM format file to filter"
+        bitwise_filter: "A set of 4 possible read filters to apply. This is a `FlagFilter` object (see ../data_structures/flag_filter.wdl for more information)."
+        prefix: "Prefix for the filtered BAM file. The extension `.bam` will be added."
+        use_all_cores: {
+            description: "Use all cores? Recommended for cloud environments.",
+            common: true
+        }
+        ncpu: {
+            description: "Number of cores to allocate for task",
+            common: true
+        }
+        modify_disk_size_gb: "Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB."
+    }
+
+    input {
+        File bam
+        FlagFilter bitwise_filter
+        String prefix = basename(bam, ".bam") + ".filtered"
+        Boolean use_all_cores = false
+        Int ncpu = 2
+        Int modify_disk_size_gb = 0
+    }
+
+    Float bam_size = size(bam, "GiB")
+    Int disk_size_gb = ceil(bam_size * 2) + 10 + modify_disk_size_gb
+
+    command <<<
+        set -euo pipefail
+
+        n_cores=~{ncpu}
+        if ~{use_all_cores}; then
+            n_cores=$(nproc)
+        fi
+        # -1 because samtools uses one more core than `--threads` specifies
+        let "n_cores -= 1"
+
+        samtools view \
+            --threads "$n_cores" \
+            -hb \
+            -f ~{bitwise_filter.include_if_all} \
+            -F ~{bitwise_filter.exclude_if_any} \
+            --rf ~{bitwise_filter.include_if_any} \
+            -G ~{bitwise_filter.exclude_if_all} \
+            ~{bam} \
+            > ~{prefix}.bam
+
+        samtools head \
+            --threads "$n_cores" \
+            -h 0 \
+            -n 1 \
+            ~{prefix}.bam \
+            > first_read.sam
+
+        if [ ! -s first_read.sam ]; then
+            >&2 echo "No reads are in the output BAM!"
+            >&2 echo "Please check that the input BAM is not empty"
+            >&2 echo "and that the filter is not too restrictive."
+            >&2 echo "Command failed!"
+            rm first_read.sam
+            exit 42
+        fi
+        rm first_read.sam
+        # TODO should there be a check that the output
+        # BAM is different from the input?
+        # It would only need to be a simple `wc -l` check.
+        # Ideally this would be done _without_ an extra pass through the BAM.
+        # An easier check is that the supplied filters aren't all zeroes.
+        # However it's still possible to have an active filter
+        # that doesn't catch anything. What's the ideal
+        # behavior in that case? The user provided something valid,
+        # so failing seems wrong. But (essentially) duplicating the BAM is also
+        # (most likely) wrong.
```

Comment:
I don't think we should add this check. I can see a use case where someone wants to drop all `duplicate` reads (or `supplemental` alignments), but not every BAM their pipeline ingests will have them. In that case, the task might often produce an identical BAM.